### DISCUSSION TITLE
Error extracting vertices from lines

### DISCUSSION THREAD

**Discussion by ccastelblanco** [cgre-aachen/gemgis#291] [Q&A]
Hello,

I would like to have your help please with an error I am experiencing when I try to use the function extract_xyz. I am trying to get the coordinates of some fault data to drape the lines over a DEM. After passing the faults gdf and the dem raster to the function extract_xyz like "gg.vector.extract_xyz(gdf=faults, dem=dem, reset_index=False)", the following error appears:

`---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\pandas\core\indexes\base.py:3653](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/pandas/core/indexes/base.py:3653), in Index.get_loc(self, key)
   3652 try:
-> 3653     return self._engine.get_loc(casted_key)
   3654 except KeyError as err:

File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\pandas\_libs\index.pyx:147](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/pandas/_libs/index.pyx:147), in pandas._libs.index.IndexEngine.get_loc()

File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\pandas\_libs\index.pyx:176](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/pandas/_libs/index.pyx:176), in pandas._libs.index.IndexEngine.get_loc()

File pandas\_libs\hashtable_class_helper.pxi:2606, in pandas._libs.hashtable.Int64HashTable.get_item()

File pandas\_libs\hashtable_class_helper.pxi:2630, in pandas._libs.hashtable.Int64HashTable.get_item()

KeyError: 1

The above exception was the direct cause of the following exception:

KeyError                                  Traceback (most recent call last)
Cell In[63], line 2
      1 faults = gdf_f_clip[gdf_f_clip.is_valid]
----> 2 faults_xyz = gg.vector.extract_xyz(gdf=faults, dem=dem, reset_index=False)

File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\gemgis\vector.py:1940](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/gemgis/vector.py:1940), in extract_xyz(gdf, dem, minz, maxz, extent, reset_index, drop_index, drop_id, drop_points, drop_level0, drop_level1, target_crs, bbox, remove_total_bounds, threshold_bounds)
   1938 # Extracting xyz
   1939 if isinstance(dem, rasterio.io.DatasetReader):
-> 1940     gdf = extract_xyz_rasterio(gdf=gdf,
   1941                                dem=dem,
   1942                                reset_index=False,
   1943                                drop_id=False,
   1944                                drop_index=False,
   1945                                drop_level0=False,
   1946                                drop_level1=False,
   1947                                drop_points=False,
   1948                                remove_total_bounds=remove_total_bounds,
   1949                                threshold_bounds=threshold_bounds)
   1951 elif isinstance(dem, np.ndarray):
   1952     gdf = extract_xyz_array(gdf=gdf,
   1953                             dem=dem,
   1954                             extent=extent,
   (...)
   1961                             remove_total_bounds=remove_total_bounds,
   1962                             threshold_bounds=threshold_bounds)

File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\gemgis\vector.py:1298](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/gemgis/vector.py:1298), in extract_xyz_rasterio(gdf, dem, minz, maxz, reset_index, drop_index, drop_id, drop_points, drop_level0, drop_level1, target_crs, bbox, remove_total_bounds, threshold_bounds)
   1296 # Extracting X and Y coordinates if they are not present in the GeoDataFrame
   1297 if not {'X', 'Y'}.issubset(gdf.columns):
-> 1298     gdf = extract_xy(gdf=gdf,
   1299                      reset_index=False,
   1300                      drop_index=False,
   1301                      drop_id=False,
   1302                      drop_points=False,
   1303                      drop_level0=False,
   1304                      drop_level1=False,
   1305                      overwrite_xy=False,
   1306                      target_crs=None,
   1307                      bbox=None,
   1308                      remove_total_bounds=remove_total_bounds,
   1309                      threshold_bounds=threshold_bounds)
   1311 # If the CRS of the gdf and the dem are identical, just extract the heights using the rasterio sample method
   1312 # NB: for points outside the bounds of the raster, nodata values will be returned
   1313 if gdf.crs == dem.crs:

File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\gemgis\vector.py:744](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/gemgis/vector.py:744), in extract_xy(gdf, reset_index, drop_index, drop_id, drop_points, drop_level0, drop_level1, overwrite_xy, target_crs, bbox, remove_total_bounds, threshold_bounds)
    742 # Extracting x,y coordinates from line vector data
    743 if all(gdf.geom_type == "LineString"):
--> 744     gdf = extract_xy_linestrings(gdf=gdf,
    745                                  reset_index=False,
    746                                  drop_id=False,
    747                                  drop_index=False,
    748                                  drop_points=False,
    749                                  overwrite_xy=overwrite_xy,
    750                                  target_crs=crs,
    751                                  bbox=bbox)
    753 # Extracting x,y coordinates from point vector data
    754 elif all(gdf.geom_type == "Point"):

File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\gemgis\vector.py:480](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/gemgis/vector.py:480), in extract_xy_linestrings(gdf, reset_index, drop_id, drop_index, drop_points, drop_level0, drop_level1, overwrite_xy, target_crs, bbox)
    477     gdf['points'] = [shapely.get_coordinates(geometry=gdf.geometry[i],
    478                                              include_z=True) for i in range(len(gdf))]
    479 else:
--> 480     gdf['points'] = [shapely.get_coordinates(geometry=gdf.geometry[i],
    481                                              include_z=False) for i in range(len(gdf))]
    483 # Creating DataFrame from exploded columns
    484 df = pd.DataFrame(data=gdf).explode('points')

File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\gemgis\vector.py:480](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/gemgis/vector.py:480), in (.0)
    477     gdf['points'] = [shapely.get_coordinates(geometry=gdf.geometry[i],
    478                                              include_z=True) for i in range(len(gdf))]
    479 else:
--> 480     gdf['points'] = [shapely.get_coordinates(geometry=gdf.geometry[i],
    481                                              include_z=False) for i in range(len(gdf))]
    483 # Creating DataFrame from exploded columns
    484 df = pd.DataFrame(data=gdf).explode('points')

File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\geopandas\geoseries.py:629](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/geopandas/geoseries.py:629), in GeoSeries.__getitem__(self, key)
    628 def __getitem__(self, key):
--> 629     return self._wrapped_pandas_method("__getitem__", key)

File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\geopandas\geoseries.py:622](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/geopandas/geoseries.py:622), in GeoSeries._wrapped_pandas_method(self, mtd, *args, **kwargs)
    620 def _wrapped_pandas_method(self, mtd, *args, **kwargs):
    621     """Wrap a generic pandas method to ensure it returns a GeoSeries"""
--> 622     val = getattr(super(), mtd)(*args, **kwargs)
    623     if type(val) == Series:
    624         val.__class__ = GeoSeries

File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\pandas\core\series.py:1007](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/pandas/core/series.py:1007), in Series.__getitem__(self, key)
   1004     return self._values[key]
   1006 elif key_is_scalar:
-> 1007     return self._get_value(key)
   1009 if is_hashable(key):
   1010     # Otherwise index.get_value will raise InvalidIndexError
   1011     try:
   1012         # For labels that don't resolve as scalars like tuples and frozensets

File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\pandas\core\series.py:1116](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/pandas/core/series.py:1116), in Series._get_value(self, label, takeable)
   1113     return self._values[label]
   1115 # Similar to Index.get_value, but we do not fall back to positional
-> 1116 loc = self.index.get_loc(label)
   1118 if is_integer(loc):
   1119     return self._values[loc]

File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\pandas\core\indexes\multi.py:2812](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/pandas/core/indexes/multi.py:2812), in MultiIndex.get_loc(self, key)
   2809     return mask
   2811 if not isinstance(key, tuple):
-> 2812     loc = self._get_level_indexer(key, level=0)
   2813     return _maybe_to_slice(loc)
   2815 keylen = len(key)

File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\pandas\core\indexes\multi.py:3160](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/pandas/core/indexes/multi.py:3160), in MultiIndex._get_level_indexer(self, key, level, indexer)
   3157         return slice(i, j, step)
   3159 else:
-> 3160     idx = self._get_loc_single_level_index(level_index, key)
   3162     if level > 0 or self._lexsort_depth == 0:
   3163         # Desired level is not sorted
   3164         if isinstance(idx, slice):
   3165             # test_get_loc_partial_timestamp_multiindex

File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\pandas\core\indexes\multi.py:2752](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/pandas/core/indexes/multi.py:2752), in MultiIndex._get_loc_single_level_index(self, level_index, key)
   2750     return -1
   2751 else:
-> 2752     return level_index.get_loc(key)

File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\pandas\core\indexes\base.py:3655](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/pandas/core/indexes/base.py:3655), in Index.get_loc(self, key)
   3653     return self._engine.get_loc(casted_key)
   3654 except KeyError as err:
-> 3655     raise KeyError(key) from err
   3656 except TypeError:
   3657     # If we have a listlike key, _check_indexing_error will raise
   3658     #  InvalidIndexError. Otherwise we fall through and re-raise
   3659     #  the TypeError.
   3660     self._check_indexing_error(key)

KeyError: 1`

Additionally, I tried to run the related cell in the example "42_draping_linestrings_over_dem_in_pyvista.ipynb" but an error results as well.

I appreciate your kind help to fix this.
 
Kind regards,

Carlos

**Comment by AlexanderJuestel** [comment]
Hello @ccastelblanco,

Would it be possible for you to provide the data for us? Then I can check directly where the error is :)

Cheers
Alex

**Comment by ccastelblanco** [comment]
Thank you! Yes, sure. Here I attach the files:

[test_files.zip](https://github.com/cgre-aachen/gemgis/files/12244542/test_files.zip)

I appreciate your help!

**Comment by AlexanderJuestel** [comment]
Hey @ccastelblanco, 

I think the error was that you also have MultiLineStrings in your data. That is why GemGIS could not handle it. We currently support only GeoDataFrames/Shape Files with one geometry type. A solution that worked for me now is the following. Here, the MultiLineStrings are exploded into single LineStrings and the index was reset. 

```python
faults = faults.explode().reset_index(drop=True)
faults_xyz = gg.vector.extract_xyz(faults, dem)
```

Let me know if that worked. I will also convert your issue to a discussion :)

**Reply by ccastelblanco** [reply]
Thank you! Yes, it worked and now I got the xyz points of the vertices. But, when I try now to create the linestrings with gg.vector.create_linestrings_from_xyz_points(gdf=faults_xyz, groupby='OBJECTID_1'), I get another error message:

`---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
Cell In[38], line 1
----> 1 faults_lines_z = gg.vector.create_linestrings_from_xyz_points(gdf=faults_xyz, groupby='OBJECTID_1')
      2 faults_lines_z

File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\gemgis\vector.py:2897](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/gemgis/vector.py:2897), in create_linestrings_from_xyz_points(gdf, groupby, nodata, xcol, ycol, zcol, dem, extent, return_gdf)
   2893 if return_gdf:
   2894     list_lines = [gpd.GeoDataFrame(
   2895         data=pd.DataFrame(data=list_gdfs_new[i].tail(1).drop(['geometry', xcol, ycol, zcol], axis=1)),
   2896         geometry=[list_linestrings_new[i]]) for i in range(len(list_linestrings_new))]
-> 2897     list_linestrings = pd.concat(list_lines).reset_index().drop(['level_0', 'level_1'], axis=1)
   2899 return list_linestrings

File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\pandas\core\frame.py:5258](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/pandas/core/frame.py:5258), in DataFrame.drop(self, labels, axis, index, columns, level, inplace, errors)
   5110 def drop(
   5111     self,
   5112     labels: IndexLabel = None,
   (...)
   5119     errors: IgnoreRaise = "raise",
   5120 ) -> DataFrame | None:
   5121     """
   5122     Drop specified labels from rows or columns.
   5123 
   (...)
   5256             weight  1.0     0.8
   5257     """
-> 5258     return super().drop(
   5259         labels=labels,
   5260         axis=axis,
   5261         index=index,
   5262         columns=columns,
   5263         level=level,
   5264         inplace=inplace,
   5265         errors=errors,
   5266     )

File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\pandas\core\generic.py:4549](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/pandas/core/generic.py:4549), in NDFrame.drop(self, labels, axis, index, columns, level, inplace, errors)
   4547 for axis, labels in axes.items():
   4548     if labels is not None:
-> 4549         obj = obj._drop_axis(labels, axis, level=level, errors=errors)
   4551 if inplace:
   4552     self._update_inplace(obj)

File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\pandas\core\generic.py:4591](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/pandas/core/generic.py:4591), in NDFrame._drop_axis(self, labels, axis, level, errors, only_slice)
   4589         new_axis = axis.drop(labels, level=level, errors=errors)
   4590     else:
-> 4591         new_axis = axis.drop(labels, errors=errors)
   4592     indexer = axis.get_indexer(new_axis)
   4594 # Case for non-unique axis
   4595 else:

File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\pandas\core\indexes\base.py:6699](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/pandas/core/indexes/base.py:6699), in Index.drop(self, labels, errors)
   6697 if mask.any():
   6698     if errors != "ignore":
-> 6699         raise KeyError(f"{list(labels[mask])} not found in axis")
   6700     indexer = indexer[~mask]
   6701 return self.delete(indexer)

KeyError: "['level_0', 'level_1'] not found in axis"`

Thank you for your kind help.

**Reply by AlexanderJuestel** [reply]
Oh no, I'll have a look at it tomorrow!

**Reply by AlexanderJuestel** [reply]
@ccastelblanco 
A hotfix that worked for me now is to add the missing columns manually using. Then the other function should work. Nevertheless, I need to fix that issue in GemGIS. Thanks for reporting that!

```python
faults_xyz['level_0'] = 0
faults_xyz['level_1'] = 0
```

**Reply by ccastelblanco** [reply]
Thank you!! Yes, it worked with that fix. But now, the function 'create_lines_3d' with 'gg.visualization.create_lines_3d(gdf=faults_lines_z)' does not work for displaying them with PyVista.

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[7], line 1
----> 1 faults_pv = gg.visualization.create_lines_3d(gdf=faults_lines_z)
      2 faults_pv

AttributeError: module 'gemgis.visualization' has no attribute 'create_lines_3d'


I appreciate your help!

**Reply by AlexanderJuestel** [reply]
I changed the name of the function a while ago 🙈 I'll get back to you with the proper name tomorrow, sorry for all that trouble

**Reply by AlexanderJuestel** [reply]
Try `create_lines_3d_polydata`

**Reply by ccastelblanco** [reply]
Yes, I thought that the name of the function probably was changed and tried with `gg.visualization.create_lines_3d_polydata(gdf=faults_lines_z)` but now got another error:

'---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[29], line 1
----> 1 faults_pv = gg.visualization.create_lines_3d_polydata(gdf=faults_lines_z)

File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\gemgis\visualization.py:115](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/gemgis/visualization.py:115), in create_lines_3d_polydata(gdf)
    112     vertices_list = vertices_list_z
    114 # Creating array of points
--> 115 points = np.vstack(vertices_list)
    117 # Creating list with number of vertices per points and indices per lines
    118 lines = []

File [c:\Users\carli\anaconda3\envs\gemgis\lib\site-packages\numpy\core\shape_base.py:289](file:///C:/Users/carli/anaconda3/envs/gemgis/lib/site-packages/numpy/core/shape_base.py:289), in vstack(tup, dtype, casting)
    287 if not isinstance(arrs, list):
    288     arrs = [arrs]
--> 289 return _nx.concatenate(arrs, 0, dtype=dtype, casting=casting)

ValueError: all the input array dimensions except for the concatenation axis must match exactly, but along dimension 1, the array at index 0 has size 3 and the array at index 96 has size 2'

Thank you again!

**Reply by AlexanderJuestel** [reply]
Alright, I can reproduce the error with your input data. Let me see what I can do now. Maybe you have found another bug ;)

**Reply by AlexanderJuestel** [reply]
Your fault at index 96 has no Z values. I continue checking
![image](https://github.com/cgre-aachen/gemgis/assets/45469915/7cfa1ce0-5ac9-42ba-8289-1f3534666a48)

**Reply by AlexanderJuestel** [reply]
It seems like the previous method did not extract any Z values for the fault (red line in the image). I would suggest adding an argument to the function that extracts the Z coordinates to drop points that are outside the dem and still extract the values for values within the dem. 

I will play around with that later. But nice use case! This is what we need to make GemGIS better! 

![image](https://github.com/cgre-aachen/gemgis/assets/45469915/c026c55e-3520-406c-a76f-5da2031675f1)

**Reply by AlexanderJuestel** [reply]
What quickly fixes your issue but removes the faults without Z coordinate is the following.

```python
faults_lines_z = faults_lines_z[faults_lines_z.has_z].reset_index(drop=True)
```

**Reply by AlexanderJuestel** [reply]
@ccastelblanco I think I tracked down the line of code that is causing the issue. However, it would be too much of a workaround to post it here. Are you okay with the solution above for now? Then I would fix things in the code and it would be available in the next version!

**Reply by AlexanderJuestel** [reply]
A fix for your issue was implemented in https://github.com/cgre-aachen/gemgis/pull/286/commits/a0a81d3554e9079bfaa15220fbc3da73d7e2b845 and will be released with version 1.1 of GemGIS :)

**Reply by AlexanderJuestel** [reply]
I also fixed tutorial 42 that you were using and fixed an error while extracting coordinates :)

**Reply by ccastelblanco** [reply]
Thank you very much! Yes, finally it works.

I appreciate your time helping me with this.

Regards,

**Reply by AlexanderJuestel** [reply]
No worries, problems of others only make our package better! So, I hope to have you back here soon!
